File: src/services/templates/langchain_api_template.py
"""
LangChain API Template - Template Component
Page-level object that places components into a layout
Articulates the design's underlying content structure
"""

from typing import Dict, List, Optional, Any
from dataclasses import dataclass
from datetime import datetime, timezone
from pydantic import BaseModel, Field

from ..organisms.langchain_integration_service import (
    LangChainIntegrationService,
    UserIntegration,
    IntegrationMetrics,
    LangChainConfig
)
from ..molecules.auth_flow_manager import AuthFlowConfig
from ..molecules.agent_connection_manager import AgentManagementConfig


class ApiTemplateContext(BaseModel):
    """API template context model"""
    user_id: str
    request_id: str
    timestamp: str
    user_agent: Optional[str] = None
    ip_address: Optional[str] = None


class ApiResponse(BaseModel):
    """Standard API response model"""
    success: bool
    data: Optional[Any] = None
    error: Optional[str] = None
    metadata: Optional[Dict[str, Any]] = None


class AuthTemplateData(BaseModel):
    """Authentication template data model"""
    providers: List[str] = Field(default_factory=list)
    auth_url: Optional[str] = None
    status: str
    token: Optional[str] = None


class AgentTemplateData(BaseModel):
    """Agent template data model"""
    agent_id: str
    name: str
    connections: int
    deployments: int
    status: str


class LangChainApiTemplateError(Exception):
    """Custom exception for API template errors"""
    pass


class LangChainApiTemplate:
    """
    Template-level API service for LangChain integration
    Places components into a coherent API structure
    Focuses on underlying content structure rather than final content
    """
    
    def __init__(self, integration_service: Optional[LangChainIntegrationService] = None):
        self.integration_service = integration_service or LangChainIntegrationService()
    
    async def handle_auth_template(
        self,
        context: ApiTemplateContext,
        provider: str,
        scopes: Optional[List[str]] = None
    ) -> ApiResponse:
        """
        Template for authentication endpoints
        Places auth components into a structured API response
        """
        start_time = datetime.now(timezone.utc)
        
        try:
            auth_config = AuthFlowConfig(
                provider=provider,
                scopes=scopes or self.integration_service.get_config().default_scopes,
                user_id=context.user_id,
            )
            
            # This would use the auth flow manager
            # For now, we'll simulate the response
            result = {
                "providers": [provider],
                "status": "pending",
                "auth_url": f"https://auth.langchain.com/oauth/{provider}",
            }
            
            response_data = AuthTemplateData(**result)
            
            return ApiResponse(
                success=True,
                data=response_data.model_dump(),
                metadata=self._create_metadata(context, start_time),
            )
            
        except Exception as e:
            return ApiResponse(
                success=False,
                error=str(e),
                metadata=self._create_metadata(context, start_time),
            )
    
    async def handle_agent_template(
        self,
        context: ApiTemplateContext,
        action: str,
        agent_id: Optional[str] = None,
        config: Optional[Dict[str, Any]] = None
    ) -> ApiResponse:
        """
        Template for agent management endpoints
        Places agent components into a structured API response
        """
        start_time = datetime.now(timezone.utc)
        
        try:
            response_data: Any = None
            
            if action == "create":
                if not config:
                    raise LangChainApiTemplateError("Configuration required for agent creation")
                
                agent_config = AgentManagementConfig(
                    name=config.get("name", "Unnamed Agent"),
                    description=config.get("description"),
                    agent_id=config.get("agent_id", self._generate_agent_id()),
                    connections=config.get("connections", []),
                    deployment_config=config.get("deployment_config", {}),
                )
                
                result = await self.integration_service.setup_complete_agent(agent_config)
                response_data = result
                
            elif action == "list":
                deployments = await self.integration_service.get_integration_metrics()
                response_data = {
                    "agents": [
                        {
                            "agent_id": f"agent_{i}",
                            "name": f"Agent {i}",
                            "connections": 0,
                            "deployments": 1,
                            "status": "active",
                        }
                        for i in range(deployments.total_deployments)
                    ]
                }
                
            elif action == "status":
                if not agent_id:
                    raise LangChainApiTemplateError("Agent ID required for status check")
                
                # This would get actual agent status
                response_data = {
                    "agent_id": agent_id,
                    "name": f"Agent {agent_id}",
                    "connections": 0,
                    "deployments": 1,
                    "status": "active",
                }
                
            elif action == "delete":
                if not agent_id:
                    raise LangChainApiTemplateError("Agent ID required for deletion")
                
                # This would actually delete the agent
                response_data = {"deleted": True}
                
            else:
                raise LangChainApiTemplateError(f"Unknown agent action: {action}")
            
            return ApiResponse(
                success=True,
                data=response_data,
                metadata=self._create_metadata(context, start_time),
            )
            
        except Exception as e:
            return ApiResponse(
                success=False,
                error=str(e),
                metadata=self._create_metadata(context, start_time),
            )
    
    async def handle_integration_template(
        self,
        context: ApiTemplateContext,
        action: str
    ) -> ApiResponse:
        """
        Template for integration status endpoints
        Places integration components into a structured API response
        """
        start_time = datetime.now(timezone.utc)
        
        try:
            response_data: Any = None
            
            if action == "status":
                integration = await self.integration_service.initialize_user_integration(
                    context.user_id,
                    ["google", "github", "microsoft"]  # Default providers
                )
                response_data = integration.model_dump()
                
            elif action == "sync":
                sync_result = await self.integration_service.sync_user_integrations(context.user_id)
                response_data = sync_result
                
            elif action == "metrics":
                metrics = await self.integration_service.get_integration_metrics(context.user_id)
                response_data = metrics.model_dump()
                
            elif action == "health":
                health = await self.integration_service.perform_health_check()
                response_data = health.model_dump()
                
            elif action == "summary":
                summary = await self.integration_service.get_user_integration_summary(context.user_id)
                response_data = summary
                
            else:
                raise LangChainApiTemplateError(f"Unknown integration action: {action}")
            
            return ApiResponse(
                success=True,
                data=response_data,
                metadata=self._create_metadata(context, start_time),
            )
            
        except Exception as e:
            return ApiResponse(
                success=False,
                error=str(e),
                metadata=self._create_metadata(context, start_time),
            )
    
    async def handle_callback_template(
        self,
        context: ApiTemplateContext,
        provider: str,
        auth_params: Dict[str, Any]
    ) -> ApiResponse:
        """
        Template for callback handling
        Places callback components into a structured response
        """
        start_time = datetime.now(timezone.utc)
        
        try:
            # Handle OAuth callback
            auth_id = auth_params.get("auth_id")
            
            if not auth_id:
                raise LangChainApiTemplateError("Auth ID required for callback")
            
            # This would use the auth flow manager to wait for completion
            # For now, we'll simulate the response
            response_data = {
                "provider": provider,
                "status": "completed",
                "success": True,
                "token": "demo-token",
                "error": None,
            }
            
            return ApiResponse(
                success=True,
                data=response_data,
                metadata=self._create_metadata(context, start_time),
            )
            
        except Exception as e:
            return ApiResponse(
                success=False,
                error=str(e),
                metadata=self._create_metadata(context, start_time),
            )
    
    async def handle_settings_template(
        self,
        context: ApiTemplateContext,
        action: str,
        config: Optional[Dict[str, Any]] = None
    ) -> ApiResponse:
        """
        Template for settings endpoints
        Places settings components into a structured response
        """
        start_time = datetime.now(timezone.utc)
        
        try:
            response_data: Any = None
            
            if action == "get":
                current_config = self.integration_service.get_config()
                response_data = {
                    "api_url": current_config.api_url,
                    "default_scopes": current_config.default_scopes,
                    "timeout_ms": current_config.timeout_ms,
                    # Don't expose the API key in the response
                }
                
            elif action == "update":
                if not config:
                    raise LangChainApiTemplateError("Configuration required for update")
                
                # Filter out sensitive fields
                safe_config = {k: v for k, v in config.items() if k != "api_key"}
                self.integration_service.update_config(safe_config)
                
                response_data = {"updated": True, "config": safe_config}
                
            else:
                raise LangChainApiTemplateError(f"Unknown settings action: {action}")
            
            return ApiResponse(
                success=True,
                data=response_data,
                metadata=self._create_metadata(context, start_time),
            )
            
        except Exception as e:
            return ApiResponse(
                success=False,
                error=str(e),
                metadata=self._create_metadata(context, start_time),
            )
    
    def create_context(
        self,
        user_id: str,
        request_id: Optional[str] = None,
        user_agent: Optional[str] = None,
        ip_address: Optional[str] = None
    ) -> ApiTemplateContext:
        """
        Create API context from request parameters
        Template helper: extracts context for all templates
        """
        return ApiTemplateContext(
            user_id=user_id,
            request_id=request_id or self._generate_request_id(),
            timestamp=datetime.now(timezone.utc).isoformat(),
            user_agent=user_agent,
            ip_address=ip_address,
        )
    
    def _create_metadata(
        self,
        context: ApiTemplateContext,
        start_time: datetime
    ) -> Dict[str, Any]:
        """
        Create standardized metadata for responses
        Template helper: creates consistent metadata structure
        """
        return {
            "request_id": context.request_id,
            "timestamp": context.timestamp,
            "processing_time_ms": int((datetime.now(timezone.utc) - start_time).total_seconds() * 1000),
            "user_id": context.user_id,
        }
    
    def _generate_request_id(self) -> str:
        """
        Generate unique request ID
        Template helper: generates request identifiers
        """
        import uuid
        return f"req_{datetime.now(timezone.utc).strftime('%Y%m%d_%H%M%S')}_{uuid.uuid4().hex[:8]}"
    
    def _generate_agent_id(self) -> str:
        """
        Generate unique agent ID
        Template helper: generates agent identifiers
        """
        import uuid
        return f"agent_{datetime.now(timezone.utc).strftime('%Y%m%d_%H%M%S')}_{uuid.uuid4().hex[:8]}"


# Singleton instance
langchain_api_template = LangChainApiTemplate()

File: src/services/templates/LangChainApiTemplate.ts
/**
 * LangChainApiTemplate - Template Component
 * Page-level object that places components into a layout
 * Articulates the design's underlying content structure
 */

import { Request, Response } from 'express';
import { langChainIntegrationService, UserIntegration, IntegrationMetrics } from '../organisms/LangChainIntegrationService';
import { authFlowManager, AuthFlowConfig } from '../molecules/AuthFlowManager';
import { agentConnectionManager, AgentManagementConfig } from '../molecules/AgentConnectionManager';

export interface ApiTemplateContext {
  userId: string;
  requestId: string;
  timestamp: string;
  userAgent?: string;
  ipAddress?: string;
}

export interface ApiResponse<T = any> {
  success: boolean;
  data?: T;
  error?: string;
  metadata?: {
    requestId: string;
    timestamp: string;
    processingTime: number;
  };
}

export interface AuthTemplateData {
  providers: string[];
  authUrl?: string;
  status: string;
  token?: string;
}

export interface AgentTemplateData {
  agentId: string;
  name: string;
  connections: number;
  deployments: number;
  status: string;
}

/**
 * Template-level API service for LangChain integration
 * Places components into a coherent API structure
 * Focuses on underlying content structure rather than final content
 */
export class LangChainApiTemplate {
  private readonly integrationService: typeof langChainIntegrationService;

  constructor() {
    this.integrationService = langChainIntegrationService;
  }

  /**
   * Template for authentication endpoints
   * Places auth components into a structured API response
   */
  async handleAuthTemplate(
    req: Request,
    res: Response,
    provider: string,
    scopes: string[] = []
  ): Promise<void> {
    const startTime = Date.now();
    const context = this.createContext(req);

    try {
      const authConfig: AuthFlowConfig = {
        provider,
        scopes: scopes.length > 0 ? scopes : this.integrationService.getConfig().defaultScopes,
        userId: context.userId,
      };

      const result = await authFlowManager.executeAuthFlow(authConfig);
      
      const responseData: AuthTemplateData = {
        providers: [provider],
        status: result.status,
        authUrl: result.authUrl,
        token: result.token,
      };

      const response = this.createApiResponse(responseData, context, startTime);
      res.json(response);
    } catch (error) {
      const response = this.createErrorResponse(error, context, startTime);
      res.status(500).json(response);
    }
  }

  /**
   * Template for agent management endpoints
   * Places agent components into a structured API response
   */
  async handleAgentTemplate(
    req: Request,
    res: Response,
    action: 'create' | 'list' | 'delete' | 'status',
    agentId?: string
  ): Promise<void> {
    const startTime = Date.now();
    const context = this.createContext(req);

    try {
      let responseData: any;

      switch (action) {
        case 'create':
          if (!req.body) {
            throw new Error('Request body required for agent creation');
          }
          const agentConfig: AgentManagementConfig = {
            name: req.body.name,
            description: req.body.description,
            agentId: req.body.agentId || this.generateAgentId(),
            connections: req.body.connections || [],
            deploymentConfig: req.body.deploymentConfig,
          };
          const createResult = await agentConnectionManager.setupAgent(agentConfig);
          responseData = createResult;
          break;

        case 'list':
          const deployments = await agentConnectionManager.getAgentDeployments();
          responseData = deployments.map(d => ({
            agentId: d.id,
            name: d.name,
            connections: 0, // Would be calculated from actual connections
            deployments: 1,
            status: d.status,
          }));
          break;

        case 'status':
          if (!agentId) {
            throw new Error('Agent ID required for status check');
          }
          const deployment = await agentConnectionManager.getAgentDeploymentStatus(agentId);
          responseData = deployment ? {
            agentId: deployment.id,
            name: deployment.name,
            connections: 0,
            deployments: 1,
            status: deployment.status,
          } : null;
          break;

        case 'delete':
          if (!agentId) {
            throw new Error('Agent ID required for deletion');
          }
          const deleteResult = await agentConnectionManager.deleteAgent(agentId, context.userId);
          responseData = { deleted: deleteResult };
          break;

        default:
          throw new Error(`Unknown agent action: ${action}`);
      }

      const response = this.createApiResponse(responseData, context, startTime);
      res.json(response);
    } catch (error) {
      const response = this.createErrorResponse(error, context, startTime);
      res.status(500).json(response);
    }
  }

  /**
   * Template for integration status endpoints
   * Places integration components into a structured API response
   */
  async handleIntegrationTemplate(
    req: Request,
    res: Response,
    action: 'status' | 'sync' | 'metrics' | 'health'
  ): Promise<void> {
    const startTime = Date.now();
    const context = this.createContext(req);

    try {
      let responseData: any;

      switch (action) {
        case 'status':
          const integration = await this.integrationService.initializeUserIntegration(
            context.userId,
            ['google', 'github', 'microsoft'] // Default providers
          );
          responseData = integration;
          break;

        case 'sync':
          const syncResult = await this.integrationService.syncUserIntegrations(context.userId);
          responseData = syncResult;
          break;

        case 'metrics':
          const metrics = await this.integrationService.getIntegrationMetrics(context.userId);
          responseData = metrics;
          break;

        case 'health':
          const health = await this.integrationService.performHealthCheck();
          responseData = health;
          break;

        default:
          throw new Error(`Unknown integration action: ${action}`);
      }

      const response = this.createApiResponse(responseData, context, startTime);
      res.json(response);
    } catch (error) {
      const response = this.createErrorResponse(error, context, startTime);
      res.status(500).json(response);
    }
  }

  /**
   * Template for callback handling
   * Places callback components into a structured response
   */
  async handleCallbackTemplate(
    req: Request,
    res: Response,
    provider: string
  ): Promise<void> {
    const startTime = Date.now();
    const context = this.createContext(req);

    try {
      // Handle OAuth callback
      const authResult = await authFlowManager.waitForAuthCompletion(
        req.query.authId as string,
        60000, // 1 minute wait
        1000   // 1 second poll interval
      );

      const responseData = {
        provider,
        status: authResult.status,
        success: authResult.success,
        token: authResult.token,
        error: authResult.error,
      };

      const response = this.createApiResponse(responseData, context, startTime);
      
      // Redirect or return based on the request type
      if (req.accepts('html')) {
        // Redirect to frontend with success/error
        const redirectUrl = authResult.success 
          ? `${process.env.FRONTEND_URL}/auth/success?provider=${provider}`
          : `${process.env.FRONTEND_URL}/auth/error?provider=${provider}&error=${authResult.error}`;
        res.redirect(redirectUrl);
      } else {
        res.json(response);
      }
    } catch (error) {
      const response = this.createErrorResponse(error, context, startTime);
      res.status(500).json(response);
    }
  }

  /**
   * Create API context from request
   * Template helper: extracts context for all templates
   */
  private createContext(req: Request): ApiTemplateContext {
    return {
      userId: req.user?.id || req.headers['x-user-id'] as string || 'anonymous',
      requestId: req.headers['x-request-id'] as string || this.generateRequestId(),
      timestamp: new Date().toISOString(),
      userAgent: req.headers['user-agent'],
      ipAddress: req.ip || req.connection.remoteAddress,
    };
  }

  /**
   * Create standardized API response
   * Template helper: creates consistent response structure
   */
  private createApiResponse<T>(
    data: T,
    context: ApiTemplateContext,
    startTime: number
  ): ApiResponse<T> {
    return {
      success: true,
      data,
      metadata: {
        requestId: context.requestId,
        timestamp: context.timestamp,
        processingTime: Date.now() - startTime,
      },
    };
  }

  /**
   * Create standardized error response
   * Template helper: creates consistent error structure
   */
  private createErrorResponse(
    error: any,
    context: ApiTemplateContext,
    startTime: number
  ): ApiResponse {
    return {
      success: false,
      error: error instanceof Error ? error.message : 'Unknown error',
      metadata: {
        requestId: context.requestId,
        timestamp: context.timestamp,
        processingTime: Date.now() - startTime,
      },
    };
  }

  /**
   * Generate unique request ID
   * Template helper: generates request identifiers
   */
  private generateRequestId(): string {
    return `req_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`;
  }

  /**
   * Generate unique agent ID
   * Template helper: generates agent identifiers
   */
  private generateAgentId(): string {
    return `agent_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`;
  }

  /**
   * Get agent deployments (helper method)
   */
  private async getAgentDeployments(): Promise<any[]> {
    // This would use the agent service to get deployments
    // For now, return empty array as placeholder
    return [];
  }
}

// Export singleton instance
export const langChainApiTemplate = new LangChainApiTemplate();
